### ISSUE TITLE
L071 produces unexpected results with array indices and `ARRAY_AGG`

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#4188]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When using array indices on array produced by `ARRAY_AGG` L071 fires and asks the user to put a white space before the array index which seems very counterintuitive. 

For example the following SQL

```sql
SELECT
    column1[SAFE_OFFSET(0)] AS column1,
    ARRAY_AGG(
        column2
        ORDER BY column2 DESC
    )[SAFE_OFFSET(0)] AS column2
FROM table1
```

### Expected Behaviour

L071 should not trigger.

### Observed Behaviour

```
L:   6 | P:   6 | L071 | Expected single whitespace between end bracket ')' and
                       | start square bracket '['.
```

### How to reproduce

Run `sqlfluff lint` on

```sql
SELECT
    column1[SAFE_OFFSET(0)] AS column1,
    ARRAY_AGG(
        column2
        ORDER BY column2 DESC
    )[SAFE_OFFSET(0)] AS column2
FROM table1
```

### Dialect

BigQuery (but probably applies to all dialects that has some function with ARRAY output)

### Version

```
sqlfluff, version 1.4.5
```

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by wong-codaio** [comment]
FWIW - A related bug caused by `L071`:
```
SELECT
  SPLIT(col, '.')[0]::INTEGER AS first_integer
FROM
  my_table
```

The rule complained that there's no space between `SPLIT(...)` and `[0]`.  In this case however, having the space actually feels awkward since it's an index-based retrieval.

**Comment by dmohns** [comment]
Another related issue, this also affect `JSON` functions. For example

```sql
SELECT PARSE_JSON('{"a": "b"}').a
```

gets corrected to

```sql
SELECT PARSE_JSON('{"a": "b"}'). a
```

**Comment by moreaupascal56** [comment]
Hi, adding my bug with snowflake parse_json

```
    parse_json(column):key
```
ges corrected to: 


```
    parse_json(column) :key
```

**Comment by kieronellis** [comment]
We're also having this issue with L071

**Comment by alanmcruickshank** [comment]
This issue from @dmohns is resolved on `main`. I believe it was resolved in #4456.

The snowflake access examples (e.g. @moreaupascal56 : `parse_json(column):key`) were resolved #4487 (also resolved #4300).

I think that's all the issues posted here so I'm going to close this one.
